packages/nextjs/app/ink/[inkId]/MintButton.tsx

```diff
@@ -2,90 +2,256 @@
+  const createMintCalls = (addresses: string[]) => {
+    return addresses.map(address => ({
+      to: NIFTY_TOKEN_CONTRACT.address,
+      data: encodeFunctionData({
+        abi: NIFTY_TOKEN_CONTRACT.abi,
+        functionName: "mint",
+        args: [address, inkId],
+      }),
+    }));
+  };
+
+  // Helper function to handle successful minting
+  const handleMintSuccess = () => {
+    notification.success("Transaction completed successfully!", {
+      icon: "🎉",
+    });
+    // Reset form and close modal on success
+    setAddresses([{ id: 1, address: "", error: "" }]);
+    setNextId(2);
+    setModalOpen(false);
+  };
+
+  // Helper function to handle single address minting fallback
+  const mintToSingleAddress = async (address: string) => {
+    await writeContractAsync({
+      abi: NIFTY_TOKEN_CONTRACT.abi,
+      address: NIFTY_TOKEN_CONTRACT.address,
+      functionName: "mint",
+      args: [address, inkId],
+    });
+    handleMintSuccess();
+  };
+
+  // Helper function to handle specific error cases
+  const handleMintError = async (error: unknown, validAddresses: string[]) => {
+    if (error instanceof Error) {
+      if (
+        error.name === "AtomicityNotSupportedError" ||
+        error.name === "AtomicReadyWalletRejectedUpgradeError" ||
+        (error.cause instanceof Error && error.cause.name === "AtomicityNotSupportedError") ||
+        (error.cause instanceof Error && error.cause.name === "AtomicReadyWalletRejectedUpgradeError")
+      ) {
+        try {
+          notification.error("Batch minting not supported. Minting only to the first address.");
+          await mintToSingleAddress(validAddresses[0]);
+          return;
+        } catch (fallbackError) {
+          console.log(fallbackError);
+        }
+      }
+    }
+
+    notification.error("Failed to mint");
+    console.log(error);
+  };
 
   const mint = async () => {
-    setMinting(true);
-    if (chain?.id !== gnosis.id) {
-      const switched = await switchTo(gnosis);
-      if (!switched) return;
+    // Step 1: Validate addresses
+    const { validAddresses, hasErrors } = validateAddresses();
+
+    if (hasErrors || validAddresses.length === 0) {
+      return;
     }
+
+    setMinting(true);
+
     try {
-      await writeContractAsync({
-        abi: NIFTY_TOKEN_CONTRACT.abi,
-        address: NIFTY_TOKEN_CONTRACT.address,
-        functionName: "mint",
-        args: [inputAddress, inkId],
-      });
-      notification.success("Transaction completed successfully!", {
-        icon: "🎉",
-      });
-      form.resetFields();
-    } catch (e) {
-      notification.error("Failed to mint");
-      console.log(e);
+      // Step 2: Ensure we're on the correct chain
+      if (chain?.id !== gnosis.id) {
+        const switched = await switchTo(gnosis);
+        if (!switched) return;
+      }
+
+      // Step 3: Execute batch mint
+      const calls = createMintCalls(validAddresses);
+      await sendCallsAsync({ calls });
```

Comment:
https://viem.sh/docs/actions/wallet/sendCalls#compatibility-fallback let's pass in the compatibility falbackhere